src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relevant flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  if (flag_bits & checksum_bitmask) {
+    frame->checksum_present = true;
+  }
+  if (flag_bits & more_to_come_bitmask) {
+    frame->more_to_come = true;
+  }
+  if (flag_bits & exhaust_allowed_bitmask) {
+    frame->exhaust_allowed = true;
+  }
+  if ((flag_bits >> 2) & required_unset_bitmask) {
+    return ParseState::kInvalid;
+  }
+
+  // Determine the number of checksum bytes in the buffer.
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
+    mongodb::Section section;
+    PX_ASSIGN_OR(section.kind, decoder->ExtractLEInt<uint8_t>(), return ParseState::kInvalid);
+    // Length of the section still remaining in the buffer.
+    int32_t remaining_section_length = 0;
+
+    if (section.kind == 0) {
+      // Check the length but don't extract it since the later logic requires the buffer to retain
+      // it.
+      section.length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
```

Comment:
Is there a maximum section length size that would indicate an invalid parse state?

Reply:
No I don't believe so, but there is a maximum document size the wire protocol mentions